authorize/sync.go

```diff
@@ -47,13 +47,17 @@ func (syncer *dataBrokerSyncer) GetDataBrokerServiceClient() databroker.DataBrok
 }
 
 func (syncer *dataBrokerSyncer) ClearRecords(ctx context.Context) {
+	syncer.authorize.stateLock.Lock()
 	syncer.authorize.store.ClearRecords()
+	syncer.authorize.stateLock.Unlock()
 }
 
 func (syncer *dataBrokerSyncer) UpdateRecords(ctx context.Context, serverVersion uint64, records []*databroker.Record) {
+	syncer.authorize.stateLock.Lock()
 	for _, record := range records {
 		syncer.authorize.store.UpdateRecord(serverVersion, record)
 	}
+	syncer.authorize.stateLock.Unlock()
```

Comment:
why not expand the interface of `store` and have UpdateRecords method there - then you do not seem to need this extra `stateLock` mutex and you won't need acquire the other mutex inside `dataBrokerData` for every `UpdateRecord()` ?

Reply:
If we remove the state lock there's no guarantee we won't upgrade records in the middle of an evaluation. You would then have an inconsistent view of the data. For example you would get record version 1234 in the audit log, but will have actually used data from version 1235 in the evaluation.

The mutex around map access is necessary because maps are not thread safe in Go. If one goroutine reads a map while another writes it it will cause the program to crash.

If having two locks in unacceptable what should I do instead to preserve thread safety and consistency?

Reply:
I missed that you also acquire RLock in `Check()`; I see why you do need that extra mutex now. 

my comment re acquiring mutex in a loop 
```go
for _, record := range records {
  syncer.authorize.store.UpdateRecord(serverVersion, record)
}
```

was about acquiring mutex on each `UpdateRecord` call, while you probably can change the interface of it to accept multiple records, and change underlying `dataBrokerData` funcs to `setLocked` and `deleteLocked` so that you could acquire mutex once outside the loop. but it's not too important probably in terms of optimization.

Reply:
This violates separation of concerns. The reason I created this type was to avoid having to think about the lock from the store, since locks are easy to misuse. If you're really concerned about performance I can make the change, but I think its negligible. FWIW during sync the method is only called with a single record.